README.md

```diff
@@ -60,7 +59,17 @@ Once you have implemented a lower bound search for the start of the range, imple
 
 *How do you use binary search to find the lower bound of a number? How did you have to modify the binary search algorithm?*
 
+# Awnser:
+I made my code so once i hit inside the interval i start scanning up/downstream (until reaching upper lower/bound) while appending features the to a new list along the way (which is then returned).
+
 *Would anything be more difficult if the features covered ranges instead of single nucleotides (like real BED files)? What could go wrong, if anything?*
 
+# Awnser:
+I guess we could have a region overlapping each end of the interval. In that case we could fix this by adding a similar 'start <= pos < end' for both ends of the interval. 
+
 *We wrote a tool for merging two BED files, but what if we had a bunch of them? What would the complexity be if we merged them in, one at a time? What would the complexity be if we merged all of the files at the same time?*
 
+# Awnser:
+Merging one at a time can be done in O(n + m) (given that they are sorted).
+That will then give us O(n + m) times the number of files (O((n+m)z) where z is the number of files). Merging all at the same time will still just require one scan so i O(n + m).
```

Comment:
This is, unfortunately, a little too optimistic. If we merged two files of size $m$ $n$ times, it would indeed take $O(nm)$ (it takes time $2m$ to merge, and we do it $n$ times). But that is not the situation if we merge the files one at a time because we have a growing file that we merge into, and each time we add another file, the next merge will be more expensive.

For simplicity, let’s say that all the files have length $m$ and that there are $n$ of them. Merging the first two would take time $2m$ (two files of length $m$). Merging in the third would take $3m$, however (one file of length $2m$ and one of length $m$). The fourth would require time $4m$ (merging one file of length $3m$ with one of length $m$).

In general, we would need the time $m(2+3+4+\cdots+n-1)$ time to merge $n$ files of length $m$ and the sum in that expression (as you know) is in $O(n^2)$, so the running time is $O(mn^2)$ (far from linear).

Concatenating all the files and sorting (with something like merge sort, as you suggest) would instead take $O(nm\log(nm))$, which could be better (depending on $n$ and $m$).

You could also, using divide and conquer, split the input in two $n/2$-length lists of files, merge these recursively, and then merge the results. That would give you the recurrence $T(n,m) = 2T(n/2,m) + O(n\cdot m)$ which is in $O(mn\log(n))$.

If you merge the $n$ files in parallel, so you, in each iteration, find the smallest front element in the $n$ files, you need to copy $nm$ elements, but it takes time to find the smallest. If you do this with a linear scan, it takes $O(n)$ to find the smallest element in each iteration, and you end up back at $O(mn^2)$. You can speed this up (once we have seen heaps in a later lecture) so it takes time $O(\log n)$ to get the smallest element per iteration, and then you are back to $O(mn\log n)$ (just with a completely different approach).

Many roads lead to Rome.